Update @joystream/types for Nicaea

Lezek123 opened the pull request:
This PR introduces the initial type definitions for the Nicaea release.
After those changes `@joystream/types` library should be compatible with the runtime based on the current `bureaucracy` branch (see: https://github.com/Joystream/joystream/pull/325).

**Circular dependency issues**
To avoid the circular dependency problem described in https://github.com/Joystream/joystream/issues/639 I decided to temporarly use the `require` statement to include the new `bureaucracy/index.ts` file at the very end of the root `/index.ts` file (which assures me that everything else is already loaded at this point). It's not a perfect solution, but until the `@joystream/types` library is restructured to avoid circular dependency issues in general, there's probably not too much that can be done here other than this or randomly changing the imports order until it starts to work.

**Error in joy-roles' `Admin.controller.tsx`** 
I also noticed an error in `joy-roles/src/tabs/Admin.controller.tsx` regarding the usage of incorrect ids in a mapping (`openingId` instead of `curatorOpeningId`). It went unnoticed before, because we've only had one module that was taking advantage of the `hiring` module, therefore those ids had always been identical, but after adding some openings via a module other than `contentWorkingGroup` (`forumBureaucracy` in that case) `workig-groups/admin` tool was not working correctly anymore until this fix was introduced.

**`types/.editorconfig`**
I also added the `types/.editorconfig` file (which was missing after moving the `types` folder away from the `pioneer` folder), because otherwise this workspace was catching the root folder `/.editorconfig` which caused VSC to incorrectly auto-indent the files inside `/types`.

**Flat namespace issues**
Due to flat namespaces in `@polkadot/api`, which have already caused us many problems, the current structure of registered types is very fragile (although it works as is). Ie. I created a type called `BureaucracyOpeningPolicyCommitment` which represents the `OpeningPolicyCommitment` from `/runtime-modules/bureaucracy/src/types.rs`, that is slightly different than the one in `/runtime-modules/content-working-group/src/lib.rs`.

Even though I declared the `policy_commitment` field of `WorkerOpening` to be of `BureaucracyOpeningPolicyCommitment` type, the `content-working-group`'s `OpeningPolicyCommitment` type is still used when sending the `forumBureaucracy.addWorkerOpening` extrinsic (which can be observed via the "Extrinsics" tab). Therefore when sending this extrinsics we have field names like `terminate_curator_role_stake_unstaking_period` instead of `terminate_worker_role_stake_unstaking_period`.

This doesn't seem to cause any bigger issues currently, probably because the fields in both those structs have exactly the same types, but it may become an issue if they start to differ in a more meaningful way.

Commit:
Update @joystream/types for Nicaea

Commit:
More clear comment

mnaamani commented on pioneer/packages/joy-roles/src/tabs/Admin.controller.tsx:
```diff
@@ -689,7 +689,7 @@ export class AdminController extends Controller<State, ITransport> {
       }
 
-      this.state.openings.set(openingId.toNumber(), {
+      this.state.openings.set(i, {
```
Nice catch. I think we caught an identical bug at one point in the runtime itself for the same exact reason.

mnaamani commented on types/.editorconfig:
```diff
@@ -0,0 +1,10 @@
+charset=utf-8
+trim_trailing_whitespace=true
+max_line_length=120
+insert_final_newline=true
```
I think we should have a common style across all the projects in the repo. @shamil-gadelshin and I briefly discussed this today. If we can pick a standard that is also matching with `cargo-fmt` for the rust code we can update just the `.editorconfig` at the root

mnaamani reviewed (changes requested):
The issue with the name conflict for type `OpeningPolicyCommitment` is unfortunate but it doesn't warrant changing names in the runtime I suppose. At some point we will replace the content working group module with an instance of a Buraucracy module so this issue will go away. 

But it might be prudent to move the identical types that are currently common to both these modules into a separate file highlighting this issue. This applies to `SlashableTerms`, `SlashingTerms` as well. Or at least add comments about above each type declaration.

Commit:
Descriptive comments about name conflicts

Commit:
Compatibility with storage_working_group_runtime_changes

Lezek123 commented:
I added a few changes to make the types compatible with the runtime after this PR: https://github.com/Joystream/joystream/pull/567 

- Replaced `BlockchainTimestamp` with `BlockAndTime`
- Added missing `StorageProviderId` type
- Added a few helper-getters to some structs (which I'm using in my later work, but since those changes are inside the `@joystream/types` workspace I dedecided to add them to this PR)
